FunctionalTableData/Array+TableSection.swift

```diff
@@ -0,0 +1,56 @@
+//
+//  Array+TableSection.swift
+//  FunctionalTableData
+//
+//  Created by Sherry Shao on 2018-06-14.
+//  Copyright © 2018 Shopify. All rights reserved.
+//
+
+extension Array where Element: TableSectionType {
+	func validateKeyUniqueness(senderName: String) {
+		let sectionKeys = map { $0.key }
+		if Set(sectionKeys).count != count {
+			let dupKeys = duplicateKeys()
+			let reason = "\(senderName) : Duplicate Section keys"
+			let userInfo: [String: Any] = ["Duplicates": dupKeys]
+			NSException(name: NSExceptionName.internalInconsistencyException, reason: reason, userInfo: userInfo).raise()
+		}
+		
+		for section in self {
+			let rowKeys = section.rows.map { $0.key }
+			if Set(rowKeys).count != section.rows.count {
+				let dupKeys = section.rows.duplicateKeys()
+				let reason = "\(senderName) : Duplicate Section.Row keys"
+				let userInfo: [String: Any] = ["Section": section.key, "Duplicates": dupKeys]
+				NSException(name: NSExceptionName.internalInconsistencyException, reason: reason, userInfo: userInfo).raise()
+			}
+		}
+	}
+}
+
+private extension Array where Element: Hashable {
+	func duplicates() -> Set<Element> {
```

Comment:
We could use https://developer.apple.com/documentation/swift/set/2853807-intersection instead

Reply:
unfortunately the actual content we care about is an array (i.e. the subject of this extension), and the duplicates in said array. 

`intersection` could only be called from a set, and while we can supply a sequence to it, calling intersect on the unique keys and supplying this array will only yield the same unique keys set.

We could also do:
```
var uniqueKeys = Set(allKeys)
return allKeys.reduce(into: Set<Int>()) {
    if uniqueKeys.contains($1) {
        uniqueKeys.remove($1)
    } else {
        $0.insert($1)
    }
}
```

I just found the current version easier to read.

Reply:
I meant internally, you are doing the exact thing intersection is doing, finding the duplicates in the sets

Reply:
So and intersection will find the common elements between two sets. What we have here is an array with duplicate elements in and of itself, if we turn it into a set, we lose that information.

Maybe I’m not understanding you correctly, do you mind giving an example?

Reply:
Oh never mind sorry, I thought you were looking for the duplicates between `sections` and `newSections`